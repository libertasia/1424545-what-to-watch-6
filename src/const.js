const DEFAULT_GENRE = `All genres`;

const MAX_GENRES_COUNT = 9;

const DEFAULT_VISIBLE_FILMS_COUNT = 8;

const NUMBER_OF_SECONDS_IN_HOUR = 3600;

const NUMBER_OF_MINUTES_IN_HOUR = 60;

const PLAYER_TOGGLER_WIDTH = 17;

const AuthorizationStatus = {
  INIT: `INIT`,
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
  AUTH_ERROR: `AUTH_ERROR`,
};

const Rating = {
  BAD: 3,
  NORMAL: 5,
  GOOD: 8,
  AWESOME: 10,
};

const RatingLevel = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`,
};

const AppRoute = {
  ADD_REVIEW: `/films/:id/review`,
  FILM: `/films/:id`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  PLAYER: `/player/:id`,
  ROOT: `/`,
};

const APIRoute = {
  ERROR_404: `/404`,
  FAVORITE_FILMS: `/favorite`,
  FILMS: `/films`,
  LOGIN: `/login`,
  LOGOUT: `/logout`,
  PROMO_FILM: `/films/promo`,
};

const TabTypes = {
  OVERVIEW: `overview`,
  DETAILS: `details`,
  REVIEWS: `reviews`,
};

const MouseEvent = {
  MOUSE_MOVE: `mousemove`,
  MOUSE_UP: `mouseup`,
};

export {DEFAULT_GENRE, MAX_GENRES_COUNT, DEFAULT_VISIBLE_FILMS_COUNT, AuthorizationStatus, Rating, RatingLevel, APIRoute, AppRoute, TabTypes, MouseEvent, NUMBER_OF_MINUTES_IN_HOUR, NUMBER_OF_SECONDS_IN_HOUR, PLAYER_TOGGLER_WIDTH};
