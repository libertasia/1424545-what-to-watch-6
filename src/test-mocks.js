const mockFilms = [
  {
    id: 1,
    name: `The Grand Budapest Hotel`,
    posterImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    previewImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    backgroundImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    backgroundColor: `#ffffff`,
    videoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
    rating: 8.9,
    scoresCount: 240,
    director: `Wes Andreson`,
    starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
    runTime: 99,
    genre: `Comedy`,
    released: 2014,
    isFavorite: true
  },
  {
    id: 2,
    name: `Stand By Me`,
    posterImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    previewImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    backgroundImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    backgroundColor: `#ffffff`,
    videoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `Rob Reiner's adaptation of Stephen King's novella The Body is a stirring, touching adventure film which knows the real world is exciting and scary enough just as it is. It's also a coming-of-age movie which celebrates the intensity of childhood friendship, while gently mourning the transience of such bonds. Which is why, unlike its central character, it'll never get old.`,
    rating: 7,
    scoresCount: 180,
    director: `Rob Reiner`,
    starring: [`River Phoenix`, `Wil Wheaton`, `Corey Feldman`, `Jerry O'Connell`],
    runTime: 89,
    genre: `Drama`,
    released: 1986,
    isFavorite: true
  },
  {
    id: 3,
    name: `Amelie`,
    posterImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    previewImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    backgroundImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
    backgroundColor: `#ffffff`,
    videoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `Despite being caught in her imaginative world, Amelie, a young waitress, decides to help people find happiness. Her quest to spread joy leads her on a journey where she finds true love.`,
    rating: 9,
    scoresCount: 355,
    director: `Jean-Pierre Jeunet`,
    starring: [`Audrey Tautou`, `Mathieu Kassovitz`, `Jamel Debbouze`],
    runTime: 120,
    genre: `Drama`,
    released: 2001,
    isFavorite: true
  },
];

const mockFilm = {
  id: 6,
  name: `The Exorcist`,
  posterImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
  previewImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
  backgroundImage: `https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto`,
  backgroundColor: `#ffffff`,
  videoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  description: `An actress notices dangerous changes in the behavior and physical make-up of her 12-year-old daughter. Meanwhile, a young priest begins to doubt his faith while dealing with his mother's sickness.`,
  rating: 8.8,
  scoresCount: 174,
  director: `William Friedkin`,
  starring: [`Linda Blair`, `Ellen Burstyn`, `Jason Miller`, `Max von Sydow`],
  runTime: 128,
  genre: `Thriller`,
  released: 1973,
  isFavorite: true
};

const mockReviews = [
  {
    id: 1,
    user: {
      id: 17,
      name: `Emely`
    },
    rating: 9.8,
    comment: `This movie is just plain bad. There must be some big payola going round this awards season. Badly written, average acting at best, all the characters are unrelatable and inlikeable. 2 hours of my life wasted.`,
    date: `2021-01-27T13:18:14.644Z`
  },
  {
    id: 2,
    user: {
      id: 13,
      name: `Zak`
    },
    rating: 8.4,
    comment: `This movie is perfect in all its categories: credits, sound track, production, casting, writing, photography, editing, acting, and direction.\nI was amazed with the freedom of the use of the camera. This movie will change the way movies are made. Slow-mo, stills, black and white, and color were all used to brilliant effect.`,
    date: `2021-01-27T13:18:14.644Z`
  },
  {
    id: 3,
    user: {
      id: 17,
      name: `Emely`
    },
    rating: 5.7,
    comment: `This movie really touched my heart, it really is the best movie of the year and everyone should see this masterpiece.`,
    date: `2021-02-06T13:18:14.644Z`
  }
];

export {mockFilms, mockFilm, mockReviews};
