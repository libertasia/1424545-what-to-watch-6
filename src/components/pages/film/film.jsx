import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import {Link, useParams} from "react-router-dom";
import PropTypes from 'prop-types';
import Tabs from './tabs';
import PageFooter from '../../shared/page-footer/page-footer';
import PageLogo from '../../shared/page-logo/page-logo';
import MovieList from '../../shared/movie-list/movie-list';
import {FilmShape, FilmsShape} from '../../../shapes';
import {fetchFilmById, fetchFilmsList, fetchReviewsById} from '../../../store/api-actions';
import {ActionCreator} from '../../../store/action';
import LoadingScreen from '../../loading-screen/loading-screen';
import UserBlock from '../../shared/user-block/user-block';
import {AuthorizationStatus} from '../../../const';
import {getAuthorizationStatus, getDataLoadedStatus, getFilm, getFilmLoadedStatus, getFilms, getReviewsLoadedStatus} from '../../../store/selectors';
import FavoriteButton from '../../shared/favorite-button/favorite-button';


const MAX_SIMILAR_FILMS_COUNT = 4;

const Film = (props) => {
  const {films, film, isDataLoaded, isFilmLoaded, isReviewsLoaded, onLoad, authorizationStatus} = props;

  const id = parseInt(useParams().id, 10);

  useEffect(() => {
    onLoad(id);
  }, [id]);

  if (!isFilmLoaded || !isReviewsLoaded || !isDataLoaded) {
    return (
      <LoadingScreen />
    );
  }

  const similarFilms = films.filter((f) => f.genre === film.genre && f.id !== film.id).slice(0, MAX_SIMILAR_FILMS_COUNT);

  const hrefToAddReviewPage = `/films/${film.id}/review`;

  const hrefToPlayerPage = `/player/${film.id}`;

  const imgAltText = `${film.name} poster`;

  return (
    <React.Fragment>
      <section className="movie-card movie-card--full">
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={film.backgroundImage} alt={film.name} />
          </div>
          <h1 className="visually-hidden">WTW</h1>
          <header className="page-header movie-card__head">
            <PageLogo />
            <UserBlock />
          </header>
          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{film.name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{film.genre}</span>
                <span className="movie-card__year">{film.released}</span>
              </p>
              <div className="movie-card__buttons">
                <Link to={hrefToPlayerPage} role="button" className="btn btn--play movie-card__button" type="button">
                  <svg viewBox="0 0 19 19" width={19} height={19}>
                    <use xlinkHref="#play-s" />
                  </svg>
                  <span>Play</span>
                </Link>
                {authorizationStatus === AuthorizationStatus.AUTH &&
                  <FavoriteButton film={film} />
                }
                {authorizationStatus === AuthorizationStatus.AUTH &&
                  <Link to={hrefToAddReviewPage} className="btn movie-card__button">Add review</Link>
                }
              </div>
            </div>
          </div>
        </div>
        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img src={film.posterImage} alt={imgAltText} width={218} height={327} />
            </div>
            <Tabs />
          </div>
        </div>
      </section>
      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>
          <MovieList visibleFilms={similarFilms} />
        </section>
        <PageFooter />
      </div>
    </React.Fragment>
  );
};

Film.propTypes = {
  authorizationStatus: PropTypes.string.isRequired,
  films: FilmsShape,
  film: FilmShape,
  isDataLoaded: PropTypes.bool.isRequired,
  isFilmLoaded: PropTypes.bool.isRequired,
  isReviewsLoaded: PropTypes.bool.isRequired,
  onLoad: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  films: getFilms(state),
  film: getFilm(state),
  isDataLoaded: getDataLoadedStatus(state),
  isFilmLoaded: getFilmLoadedStatus(state),
  isReviewsLoaded: getReviewsLoadedStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  onLoad(id) {
    dispatch(ActionCreator.resetActiveTab());
    dispatch(fetchFilmsList());
    dispatch(fetchFilmById(id));
    dispatch(fetchReviewsById(id));
  },
});

export {Film};
export default connect(mapStateToProps, mapDispatchToProps)(Film);
