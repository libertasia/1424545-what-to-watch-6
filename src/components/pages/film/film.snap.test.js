import React from 'react';
import {Provider} from 'react-redux';
import {render} from '@testing-library/react';
import {Router} from 'react-router-dom';
import configureStore from 'redux-mock-store';
import {createMemoryHistory} from 'history';
import {Film} from './film';
import {mockFilm, mockFilms, mockReviews} from '../../../test-mocks';
import {AuthorizationStatus, TabTypes} from '../../../const';

const mockStore = configureStore({});
const store = {
  FILMS: {
    authorizationStatus: AuthorizationStatus.AUTH,
    film: mockFilm,
    films: mockFilms,
    reviews: mockReviews,
    isDataLoaded: true,
    isFilmLoaded: true,
    isReviewsLoaded: true,
  },
  VIEW: {
    activeTab: TabTypes.OVERVIEW,
  }
};
it(`Film renders correctly`, () => {
  const history = createMemoryHistory();
  const {container} = render(
      <Provider store={mockStore(store)}>
        <Router history={history}>
          <Film
            film={mockFilm}
            films={mockFilms}
            isDataLoaded={true}
            isFilmLoaded={true}
            isReviewsLoaded={true}
            onLoad={jest.fn()}
            authorizationStatus={AuthorizationStatus.AUTH}
          />
        </Router>
      </Provider>
  );

  expect(container).toMatchSnapshot();
});
