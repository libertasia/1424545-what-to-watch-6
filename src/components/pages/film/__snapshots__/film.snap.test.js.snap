// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Film renders correctly 1`] = `
<div>
  <section
    class="movie-card movie-card--full"
  >
    <div
      class="movie-card__hero"
    >
      <div
        class="movie-card__bg"
      >
        <img
          alt="The Exorcist"
          src="https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto"
        />
      </div>
      <h1
        class="visually-hidden"
      >
        WTW
      </h1>
      <header
        class="page-header movie-card__head"
      >
        <div
          class="logo"
        >
          <a
            class="logo__link"
            href="/"
          >
            <span
              class="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              class="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              class="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          class="user-block"
        >
          <div
            class="user-block__avatar"
          >
            <a
              href="/mylist"
            >
              <img
                alt="User avatar"
                height="63"
                src="img/avatar.jpg"
                width="63"
              />
            </a>
          </div>
        </div>
      </header>
      <div
        class="movie-card__wrap"
      >
        <div
          class="movie-card__desc"
        >
          <h2
            class="movie-card__title"
          >
            The Exorcist
          </h2>
          <p
            class="movie-card__meta"
          >
            <span
              class="movie-card__genre"
            >
              Thriller
            </span>
            <span
              class="movie-card__year"
            >
              1973
            </span>
          </p>
          <div
            class="movie-card__buttons"
          >
            <a
              class="btn btn--play movie-card__button"
              href="/player/6"
              role="button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlink:href="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <button
              class="btn btn--list movie-card__button"
              data-testid="favorite_btn"
              type="button"
            >
              <svg
                height="14"
                viewBox="0 0 18 14"
                width="18"
              >
                <use
                  xlink:href="#in-list"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              class="btn movie-card__button"
              href="/films/6/review"
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="movie-card__wrap movie-card__translate-top"
    >
      <div
        class="movie-card__info"
      >
        <div
          class="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="The Exorcist poster"
            height="327"
            src="https://i.picsum.photos/id/604/248/152.jpg?hmac=Yu4rnPqlL8pPa8b7cdPSosfFApQXtQmy56BW52elqto"
            width="218"
          />
        </div>
        <div
          class="movie-card__desc"
        >
          <nav
            class="movie-nav movie-card__nav"
          >
            <ul
              class="movie-nav__list"
            >
              <li
                class="movie-nav__item movie-nav__item--active"
                data-id="overview"
              >
                <a
                  class="movie-nav__link"
                  href="#"
                >
                  Overview
                </a>
              </li>
              <li
                class="movie-nav__item "
                data-id="details"
              >
                <a
                  class="movie-nav__link"
                  href="#"
                >
                  Details
                </a>
              </li>
              <li
                class="movie-nav__item "
                data-id="reviews"
              >
                <a
                  class="movie-nav__link"
                  href="#"
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div
            class="movie-rating"
          >
            <div
              class="movie-rating__score"
            >
              8.8
            </div>
            <p
              class="movie-rating__meta"
            >
              <span
                class="movie-rating__level"
              >
                Very good
              </span>
              <span
                class="movie-rating__count"
              >
                174
                 ratings
              </span>
            </p>
          </div>
          <div
            class="movie-card__text"
          >
            <p>
              An actress notices dangerous changes in the behavior and physical make-up of her 12-year-old daughter. Meanwhile, a young priest begins to doubt his faith while dealing with his mother's sickness.
            </p>
            <p
              class="movie-card__director"
            >
              <strong>
                Director: 
                William Friedkin
              </strong>
            </p>
            <p
              class="movie-card__starring"
            >
              <strong>
                Starring: 
                Linda Blair, Ellen Burstyn, Jason Miller, Max von Sydow
                 and other
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    class="page-content"
  >
    <section
      class="catalog catalog--like-this"
    >
      <h2
        class="catalog__title"
      >
        More like this
      </h2>
      <div
        class="catalog__movies-list"
      />
    </section>
    <footer
      class="page-footer"
    >
      <div
        class="logo"
      >
        <a
          class="logo__link logo__link--light"
          href="/"
        >
          <span
            class="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            class="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            class="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        class="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;
